phoenix-core/src/main/java/org/apache/phoenix/index/IndexMaintainer.java

```diff
@@ -326,6 +326,49 @@ private IndexMaintainer(PTable dataTable, PTable index) {
         // TODO: check whether index is immutable or not. Currently it's always false so checking
         // data table is with immutable rows or not.
         this.immutableRows = dataTable.isImmutableRows();
+        int indexColByteSize = 0;
+        for (int i = indexPosOffset; i < index.getPKColumns().size(); i++) {
+            PColumn indexColumn = index.getPKColumns().get(i);
+            int indexPos = i - indexPosOffset;
+            Expression expression = null;
+            try {
+                ParseNode parseNode  = SQLParser.parseCondition(indexColumn.getExpressionStr());
+                ColumnResolver resolver = FromCompiler.getResolver(new TableRef(dataTable));
+                StatementContext context = new StatementContext(new PhoenixStatement(connection), resolver);
+                expression = parseNode.accept(new ExpressionCompiler(context));
```

Comment:
If the null check idea above won't work, instead of the instanceof check below, can you track in the ExpressionCompiler if you've compiled an expression versus only a column that maps to the PK column in the data table? You could specialize the resolveColumn method of your ExpressionCompiler and check visitor.isTopLevel() and SchemaUtil.isPKColumn(IndexUtil.getDataColumn(dataTable, indexColumn.getName().getString()))